File: backend/iis_client.py
"""
IIS management via WinRM (PowerShell remoting).

Falls back to TCP port-check for status when WinRM is not available.
"""
import logging
from typing import Tuple
from websphere_client import check_port

logger = logging.getLogger(__name__)

try:
    import winrm
    WINRM_AVAILABLE = True
except ImportError:
    WINRM_AVAILABLE = False
    logger.warning("pywinrm not installed – IIS management disabled. Run: pip install pywinrm")


def _get_protocol(server: dict):
    """Create a pywinrm Protocol object from server config."""
    use_ssl = server.get("winrm_use_ssl", False)
    scheme = "https" if use_ssl else "http"
    port = server.get("winrm_port", 5985)
    endpoint = f"{scheme}://{server['host']}:{port}/wsman"
    return winrm.Protocol(
        endpoint=endpoint,
        transport="ntlm",
        username=server.get("winrm_username", ""),
        password=server.get("winrm_password", ""),
        server_cert_validation="ignore",
    )


def _run_ps(server: dict, script: str, timeout: int = 60) -> Tuple[int, str, str]:
    """Run a PowerShell script via WinRM. Returns (rc, stdout, stderr)."""
    proto = _get_protocol(server)
    shell_id = proto.open_shell()
    try:
        command_id = proto.run_command(
            shell_id,
            "powershell",
            ["-NonInteractive", "-NoProfile", "-Command", script],
        )
        stdout, stderr, rc = proto.get_command_output(shell_id, command_id)
        return rc, stdout.decode(errors="replace"), stderr.decode(errors="replace")
    finally:
        proto.cleanup_command(shell_id, command_id)  # type: ignore[possibly-undefined]
        proto.close_shell(shell_id)


# ── Public API ────────────────────────────────────────────────────

def get_server_status(server: dict) -> Tuple[str, str]:
    """
    Check IIS server status. Returns (status, message).
    Tries WinRM first; falls back to TCP port 80/443.
    """
    if not WINRM_AVAILABLE:
        # Fallback: TCP port check on 80
        host = server.get("host", "")
        if check_port(host, 80):
            return "running", "Port 80 open (WinRM unavailable – limited check)"
        if check_port(host, 443):
            return "running", "Port 443 open (WinRM unavailable – limited check)"
        return "unknown", "pywinrm not installed; TCP check inconclusive"

    ps_script = """
Import-Module WebAdministration -ErrorAction SilentlyContinue
$results = @()
Get-Website | ForEach-Object { $results += "$($_.Name)=$($_.State)" }
$results -join ','
"""
    try:
        rc, stdout, stderr = _run_ps(server, ps_script)
        if rc == 0:
            output = stdout.strip()
            if "Started" in output:
                return "running", f"IIS sites: {output}"
            if "Stopped" in output:
                return "stopped", f"IIS sites: {output}"
            return "unknown", f"IIS state: {output}"
        return "unknown", f"WinRM rc={rc}: {stderr[:200]}"
    except Exception as exc:
        logger.warning("IIS WinRM check failed for %s: %s", server.get("host"), exc)
        # TCP fallback
        host = server.get("host", "")
        if check_port(host, 80) or check_port(host, 443):
            return "running", f"WinRM failed ({exc}); TCP port open"
        return "unknown", f"WinRM error: {exc}"


def start_server(server: dict) -> Tuple[bool, str]:
    """Start all configured IIS sites and app pools."""
    if not WINRM_AVAILABLE:
        return False, "pywinrm not installed – cannot start IIS remotely"

    sites = server.get("iis_sites", [])
    ps_lines = ["Import-Module WebAdministration -ErrorAction SilentlyContinue"]
    for site in sites:
        site_name = site.get("name", "Default Web Site")
        ps_lines.append(f'Start-WebSite -Name "{site_name}" -ErrorAction SilentlyContinue')
        for pool in site.get("app_pools", []):
            ps_lines.append(f'Start-WebAppPool -Name "{pool}" -ErrorAction SilentlyContinue')
    ps_lines.append('Write-Output "Done"')

    try:
        rc, stdout, stderr = _run_ps(server, "\n".join(ps_lines), timeout=60)
        if rc == 0:
            return True, f"IIS sites started on {server['host']}"
        return False, f"IIS start rc={rc}: {stderr[:300]}"
    except Exception as exc:
        logger.exception("IIS start failed for %s", server.get("host"))
        return False, f"WinRM error: {exc}"


def stop_server(server: dict) -> Tuple[bool, str]:
    """Stop all configured IIS sites and app pools."""
    if not WINRM_AVAILABLE:
        return False, "pywinrm not installed – cannot stop IIS remotely"

    sites = server.get("iis_sites", [])
    ps_lines = ["Import-Module WebAdministration -ErrorAction SilentlyContinue"]
    for site in sites:
        site_name = site.get("name", "Default Web Site")
        ps_lines.append(f'Stop-WebSite -Name "{site_name}" -ErrorAction SilentlyContinue')
        for pool in site.get("app_pools", []):
            ps_lines.append(f'Stop-WebAppPool -Name "{pool}" -ErrorAction SilentlyContinue')
    ps_lines.append('Write-Output "Done"')

    try:
        rc, stdout, stderr = _run_ps(server, "\n".join(ps_lines), timeout=60)
        if rc == 0:
            return True, f"IIS sites stopped on {server['host']}"
        return False, f"IIS stop rc={rc}: {stderr[:300]}"
    except Exception as exc:
        logger.exception("IIS stop failed for %s", server.get("host"))
        return False, f"WinRM error: {exc}"


def restart_server(server: dict) -> Tuple[bool, str]:
    """Perform an iisreset on the target server."""
    if not WINRM_AVAILABLE:
        return False, "pywinrm not installed – cannot restart IIS remotely"

    try:
        rc, stdout, stderr = _run_ps(server, "iisreset /restart", timeout=120)
        if rc == 0:
            return True, f"IIS restarted on {server['host']}"
        return False, f"iisreset rc={rc}: {stderr[:300]}"
    except Exception as exc:
        logger.exception("IIS restart failed for %s", server.get("host"))
        return False, f"WinRM error: {exc}"

File: java/src/main/java/com/ibm/was/dashboard/service/IisClient.java
package com.ibm.was.dashboard.service;

import org.apache.hc.client5.http.auth.AuthScope;
import org.apache.hc.client5.http.auth.NTCredentials;
import org.apache.hc.client5.http.classic.methods.HttpPost;
import org.apache.hc.client5.http.impl.auth.BasicCredentialsProvider;
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
import org.apache.hc.client5.http.impl.classic.HttpClientBuilder;
import org.apache.hc.client5.http.ssl.NoopHostnameVerifier;
import org.apache.hc.client5.http.ssl.SSLConnectionSocketFactoryBuilder;
import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManagerBuilder;
import org.apache.hc.core5.http.ContentType;
import org.apache.hc.core5.http.io.entity.EntityUtils;
import org.apache.hc.core5.http.io.entity.StringEntity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;
import java.security.cert.X509Certificate;
import java.util.List;
import java.util.Map;

/**
 * IIS management via WinRM PowerShell over HTTP/NTLM.
 * Falls back to TCP port check if WinRM is unavailable.
 */
@Component
public class IisClient {

    private static final Logger log = LoggerFactory.getLogger(IisClient.class);

    // ── Status ────────────────────────────────────────────────────

    public String[] getStatus(Map<String, Object> server) {
        String host = str(server, "host");
        if (host.isEmpty()) return new String[]{"unknown", "No host configured"};

        try {
            String output = runPowerShell(server,
                "Import-Module WebAdministration -ErrorAction SilentlyContinue;" +
                "Get-Website | ForEach-Object { \"$($_.Name)=$($_.State)\" }");

            if (output.contains("Started")) return new String[]{"running", "IIS: " + truncate(output, 200)};
            if (output.contains("Stopped")) return new String[]{"stopped", "IIS: " + truncate(output, 200)};
            return new String[]{"unknown", "IIS state: " + truncate(output, 200)};

        } catch (Exception e) {
            log.debug("WinRM status failed for {}: {} – falling back to TCP", host, e.getMessage());
            if (WebSphereClient.portOpen(host, 80) || WebSphereClient.portOpen(host, 443)) {
                return new String[]{"running", "WinRM unavailable; TCP port open"};
            }
            return new String[]{"unknown", "WinRM error: " + e.getMessage()};
        }
    }

    // ── Actions ───────────────────────────────────────────────────

    public String[] start(Map<String, Object> server) {
        return iisAction(server, "start");
    }

    public String[] stop(Map<String, Object> server) {
        return iisAction(server, "stop");
    }

    public String[] restart(Map<String, Object> server) {
        try {
            String out = runPowerShell(server, "iisreset /restart");
            return new String[]{"ok", "IIS restarted: " + truncate(out, 200)};
        } catch (Exception e) {
            return new String[]{"error", "IIS restart failed: " + e.getMessage()};
        }
    }

    @SuppressWarnings("unchecked")
    private String[] iisAction(Map<String, Object> server, String verb) {
        List<Map<String, Object>> sites =
            (List<Map<String, Object>>) server.getOrDefault("iis_sites", List.of());

        StringBuilder ps = new StringBuilder(
            "Import-Module WebAdministration -ErrorAction SilentlyContinue;\n");

        String cmd = verb.equals("start") ? "Start" : "Stop";
        for (Map<String, Object> site : sites) {
            String siteName = str(site, "name");
            ps.append(cmd).append("-WebSite -Name \"").append(siteName)
              .append("\" -ErrorAction SilentlyContinue;\n");

            List<String> pools = (List<String>) site.getOrDefault("app_pools", List.of());
            for (String pool : pools) {
                ps.append(cmd).append("-WebAppPool -Name \"").append(pool)
                  .append("\" -ErrorAction SilentlyContinue;\n");
            }
        }
        ps.append("Write-Output 'Done'");

        try {
            String out = runPowerShell(server, ps.toString());
            return new String[]{"ok", "IIS " + verb + " complete: " + truncate(out, 100)};
        } catch (Exception e) {
            return new String[]{"error", "IIS " + verb + " failed: " + e.getMessage()};
        }
    }

    // ── WinRM SOAP over HTTP with NTLM ─────────────────────────────

    private String runPowerShell(Map<String, Object> server, String script) throws Exception {
        String host     = str(server, "host");
        int    port     = intVal(server, "winrm_port", 5985);
        boolean useSsl  = Boolean.TRUE.equals(server.get("winrm_use_ssl"));
        String username = str(server, "winrm_username");
        String password = str(server, "winrm_password");
        String scheme   = useSsl ? "https" : "http";
        String endpoint = scheme + "://" + host + ":" + port + "/wsman";

        // Build encoded PowerShell command
        String encoded = java.util.Base64.getEncoder()
            .encodeToString(script.getBytes(java.nio.charset.StandardCharsets.UTF_16LE));

        String soapBody = buildWinRMSoapEnvelope("powershell -EncodedCommand " + encoded);

        BasicCredentialsProvider creds = new BasicCredentialsProvider();
        // Split domain\user
        String domain = "", user = username;
        if (username.contains("\\")) {
            domain = username.substring(0, username.indexOf('\\'));
            user   = username.substring(username.indexOf('\\') + 1);
        }
        creds.setCredentials(new AuthScope(host, port),
            new NTCredentials(user, password.toCharArray(), host, domain));

        try (CloseableHttpClient client = buildHttpClient(creds, useSsl)) {
            HttpPost post = new HttpPost(endpoint);
            post.setHeader("Content-Type", "application/soap+xml;charset=UTF-8");
            post.setHeader("WSMANIDENTIFY", "unauthenticated");
            post.setEntity(new StringEntity(soapBody, ContentType.create("application/soap+xml", "UTF-8")));

            return client.execute(post, response -> {
                String body = EntityUtils.toString(response.getEntity());
                if (response.getCode() >= 400) {
                    throw new RuntimeException("WinRM HTTP " + response.getCode() + ": " + truncate(body, 200));
                }
                return extractWinRMOutput(body);
            });
        }
    }

    private CloseableHttpClient buildHttpClient(BasicCredentialsProvider creds, boolean useSsl) throws Exception {
        if (useSsl) {
            SSLContext sslCtx = SSLContext.getInstance("TLS");
            sslCtx.init(null, new TrustManager[]{new X509TrustManager() {
                public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
                public void checkClientTrusted(X509Certificate[] c, String a) {}
                public void checkServerTrusted(X509Certificate[] c, String a) {}
            }}, null);

            var cm = PoolingHttpClientConnectionManagerBuilder.create()
                .setSSLSocketFactory(SSLConnectionSocketFactoryBuilder.create()
                    .setSslContext(sslCtx).setHostnameVerifier(NoopHostnameVerifier.INSTANCE).build())
                .build();

            return HttpClientBuilder.create()
                .setDefaultCredentialsProvider(creds)
                .setConnectionManager(cm).build();
        }
        return HttpClientBuilder.create()
            .setDefaultCredentialsProvider(creds).build();
    }

    private String buildWinRMSoapEnvelope(String command) {
        return """
            <s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"
                        xmlns:wsmid="http://schemas.dmtf.org/wbem/wsman/identity/1/wsmanidentity.xsd"
                        xmlns:wsman="http://schemas.dmtf.org/wbem/wsman/1/wsman.xsd"
                        xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing"
                        xmlns:rsp="http://schemas.microsoft.com/wbem/wsman/1/windows/shell">
              <s:Header>
                <wsman:ResourceURI>http://schemas.microsoft.com/wbem/wsman/1/windows/shell/cmd</wsman:ResourceURI>
                <wsa:Action>http://schemas.xmlsoap.org/ws/2004/09/transfer/Create</wsa:Action>
              </s:Header>
              <s:Body>
                <rsp:Shell><rsp:Command>""" + escapeXml(command) + """
                </rsp:Command></rsp:Shell>
              </s:Body>
            </s:Envelope>""";
    }

    private String extractWinRMOutput(String soap) {
        // Very lightweight extraction – avoids pulling in a full XML parser
        int start = soap.indexOf("<rsp:Stream");
        if (start < 0) return soap;
        int end   = soap.lastIndexOf("</rsp:Stream>") + "</rsp:Stream>".length();
        String raw = soap.substring(start, Math.min(end, soap.length()));
        // Strip XML tags
        return raw.replaceAll("<[^>]+>", "").trim();
    }

    // ── Helpers ───────────────────────────────────────────────────

    private static String str(Map<String, Object> m, String key) {
        Object v = m.get(key); return v == null ? "" : v.toString();
    }

    private static int intVal(Map<String, Object> m, String key, int def) {
        Object v = m.get(key);
        if (v instanceof Number n) return n.intValue();
        if (v instanceof String s) { try { return Integer.parseInt(s); } catch (Exception ignored) {} }
        return def;
    }

    private static String escapeXml(String s) {
        return s.replace("&","&amp;").replace("<","&lt;").replace(">","&gt;")
                .replace("\"","&quot;").replace("'","&apos;");
    }

    private static String truncate(String s, int max) {
        return s != null && s.length() > max ? s.substring(0, max) + "…" : (s == null ? "" : s);
    }
}
